### PR TITLE
Daily forecast design iteration

### PR DESCRIPTION
**Pull request by eric-gade**
## What does this PR do? 🛠️
This PR implements the design changes outlined in #680 

### Structural changes ###
Previously, in the Daily forecast we excluded any periods corresponding to the current day, and any forecast periods after the variable limit in the number of days to show.
  
This has changed.
  
Now, we essentially process the daily forecast periods into three parts:
1. **"Today" periods**: These correspond to periods that align with the current day/night when the user is looking at the forecast. Examples of period names here include "Today" "This Afternoon" "Tonight" and "Overnight". There are usually 1 or 2 of these periods, depending on when the user looks.
2. **Detailed forecast days**: These are the days where we show the `shortForecast` text (and later on in subsequent tickets, further details). We limit these days by some variable number, by default we use 3 (4 if you include the "Today" periods).
3. **"Extended"** periods**: These are essentially all the remaining periods the API returns that come _after_ the limited number of detailed forecast days we've selected
  
These changes required a refactor of the methods in `WeatherDataService` pertaining to processing daily forecast information. We have also added and updated tests where needed.
  
### Partials ###
This PR also introduces the use of twig template partials, in order to decrease repeating ourselves in larger views. The daily forecast template now makes use of two partials. You can see an example of partial inclusion, with parameters, [here](https://github.com/weather-gov/weather.gov/blob/eg-680-iter-daily/web/themes/new_weather_theme/templates/block/block--weathergov-daily-forecast.html.twig#L7). You can find the theme's [partial directory here](https://github.com/weather-gov/weather.gov/tree/eg-680-iter-daily/web/themes/new_weather_theme/templates/partials).

## Screenshots (if appropriate): 📸
<details>
<summary>Desktop:</summary>

Screen Shot 2024-01-25 at 3 52 18 PM
Screen Shot 2024-01-25 at 3 52 24 PM

</details>
  
<details>
<summary>Mobile:</summary>

Screen Shot 2024-01-25 at 3 52 40 PM
Screen Shot 2024-01-25 at 3 52 49 PM

</details>

### CODE REVIEW

File: web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#}
+<li class="display-block margin-bottom-5">
+  <div class="daily-date-and-temp display-flex flex-row flex-justify flex-align-center margin-bottom-2">
+    <time class="display-block flex-1 text-bold font-heading-md" datetime="{{ periods[0].startTime }}">
+      {% if dayLabel %}
+        <span>{{ dayLabel }},</span>
+      {% else %}
+        <span>{{ periods[0].dayName }}</span>
+      {% endif %}
+      <span class="display-block">{{ periods[0].monthAndDay }}</span>
```
Comment (igorkorenfeld) [inline web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig:18]: could you add `desktop:display-inline` to this one so it's one line on larger screens

File: web/themes/new_weather_theme/templates/partials/daily-high-low.html.twig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<div class="daily-high-low display-flex flex-column flex-align-center padding-x-4px margin-right-1">
```
Comment (igorkorenfeld) [inline web/themes/new_weather_theme/templates/partials/daily-high-low.html.twig:1]: margin on the right is a little tighter than the design, could we try `margin-right-2`

File: web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      {% include '@new_weather_theme/partials/daily-high-low.html.twig' with { 'period': period } %}
+    {% endfor %}
+  </div>
+  {% for period in periods %}
+    {% set label =  period.isDaytime ? "Day" | t : "Night" | t %}
+    <div class="daily-forecast-summary-area margin-bottom-2">
+      <span class="display-block text-bold font-body-xs">
+        {{label}}
+      </span>
+      <div class="display-flex flex-row flex-align-start">
```
Comment (igorkorenfeld) [inline web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig:31]: the alignemnt looks a little off with short lines, could we change this one to `flex-align-center`

File: web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    </time>
+
+    {% for period in periods %}
+      {% include '@new_weather_theme/partials/daily-high-low.html.twig' with { 'period': period } %}
+    {% endfor %}
+  </div>
+  {% for period in periods %}
+    {% set label =  period.isDaytime ? "Day" | t : "Night" | t %}
+    <div class="daily-forecast-summary-area margin-bottom-2">
+      <span class="display-block text-bold font-body-xs">
```
Comment (igorkorenfeld) [inline web/themes/new_weather_theme/templates/partials/daily-forecast-list-item.html.twig:28]: Need a little space under the day/night label, could you add `margin-bottom-1` here

**Review by igorkorenfeld — COMMENTED** [review]
Had a few small design edits

**Review by igorkorenfeld — COMMENTED** [review]
One more update to roll into this PR if possible:

Could you add a `margin-top-4` and `desktop:grid-col-6` on the `usa-accordion`  ?

File: web/modules/weather_data/src/Service/WeatherDataService.php [hunk 5]
```diff
@@ -153,6 +160,94 @@ protected function isValidObservation($obs)
...
+        // } else {
+        //     $tonightPeriod = $periods[1];
+        //     $todayPeriod = $periods[0];
+        // }
+
+        // $result = [];
+        // if ($todayPeriod) {
+        //     $result["daytime"] = $this->formatDailyPeriod($todayPeriod);
+        // }
+        // $result["overnight"] = $this->formatDailyPeriod($tonightPeriod);
```
Comment (greg-does-weather) [inline web/modules/weather_data/src/Service/WeatherDataService.php:205]:
```suggestion
```

Is this cruft we can remove?

File: web/modules/weather_data/src/Service/WeatherDataService.php [hunk 6]
```diff
@@ -609,51 +735,76 @@ public function getDailyForecastFromGrid(
                 $periods[0]->startTime,
             );
         }
-        $periods = $this->filterToFutureDays($periods, $now, $defaultDays);
 
-        // Periods are either daytime or nighttime
-        // We can zip them together as pairs.
-        $dayNightPairs = array_chunk($periods, 2);
+        // $periods = $this->filterToFutureDays($periods, $now, $defaultDays);
```
Comment (greg-does-weather) [inline web/modules/weather_data/src/Service/WeatherDataService.php:739]:
```suggestion
```

**Review by greg-does-weather — APPROVED** [review]
Engineering review looks good. I'll leave it to you and @igorkorenfeld to decide when it's ready to merge.

## Code Review Checklist

A checked item is either an assertion that I have tested this item or an indication that I have verified it does not apply to this pull request.

## The Basics
- [x] Checks are passing
- [x] I read the code
- [x] I ran the code
- [x] (if applicable) Post deploy steps are run
- [x] (if applicable) I validated the change on the deployed version in <environment>

## Documentation
- [x] changes to “how we do things” are documented in READMEs
- [x] all new functions and methods are commented using plain language
- [x] any new modules added documented in modules.md

## Security
- [x] security false positives are documented
- [x] data from external sources is cleaned and clearly marked

## Reliability
- [x] error handling exists for unusual or missing values
- [x] interactions with external systems are wrapped in try/except
- [x] functionality is tested with unit or integration tests
- [x] dependency updates in composer.json also got changed in composer-lock.json

## Infrastructure
- [x] all changes are auditable and documented via a script
- [x] it is clear who can and should run the script
- [x] (if applicable) diagrams have been updated or added in PlantUML

## Accessibility
- [x] New pages have been added to cypress-axe file so that they will be tested with our automated accessibility testing
- [x] Meets WCAG 2.0 AA or 2.1 AA for Section 508 compliance

## Device Matrix
- [x] firefox/gecko (renders correctly and user interactions work)
- [x] chrome/chromium/edge (renders correctly and user interactions work)
- [x] safari/webkit (renders correctly and user interactions work)
- [x] web page is readable and usable 
    - [x] at 480px     (mobile)
    - [x] at 640px     (tablet)
    - [x] at 1024px    (desktop)

**Comment by greg-does-weather** [comment]
@igorkorenfeld Personal observation, when the page width transitions from the desktop breakpoint to the tablet one, the daily forecast suddenly gets super wide.

**Comment by igorkorenfeld** [comment]
Good callout @greg-does-weather, working with Eric on how to adjust so it's a little more readable.

**Pull request merged**